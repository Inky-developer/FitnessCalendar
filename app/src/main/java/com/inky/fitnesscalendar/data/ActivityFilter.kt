package com.inky.fitnesscalendar.data

import java.time.Instant
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.ZoneId
import java.util.Date


data class ActivityFilter(
    val types: List<ActivityType> = emptyList(),
    val text: String? = null,
    val startRangeDate: Date? = null,
    val endRangeDate: Date? = null
) {
    fun isEmpty() = this == ActivityFilter()

    companion object {
        fun atDay(instant: Instant): ActivityFilter {
            val day = LocalDateTime.ofInstant(instant, ZoneId.systemDefault())
            val startOfToday = day.with(LocalTime.MIN)
            val endOfToday = day.with(LocalTime.MAX)

            return ActivityFilter(
                startRangeDate = Date.from(startOfToday.atZone(ZoneId.systemDefault()).toInstant()),
                endRangeDate = Date.from(endOfToday.atZone(ZoneId.systemDefault()).toInstant())
            )
        }
    }
}