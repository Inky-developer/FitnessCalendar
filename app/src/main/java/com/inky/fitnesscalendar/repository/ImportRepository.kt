package com.inky.fitnesscalendar.repository

import android.os.ParcelFileDescriptor
import androidx.compose.runtime.Immutable
import com.inky.fitnesscalendar.data.gpx.GpxTrack
import com.inky.fitnesscalendar.data.gpx.TrackSvg
import com.inky.fitnesscalendar.data.gpx.TrackSvg.Companion.toTrackSvg
import com.inky.fitnesscalendar.db.entities.Activity
import com.inky.fitnesscalendar.db.entities.Activity.SerializedTrackPreview.Companion.serialize
import com.inky.fitnesscalendar.db.entities.ActivityType
import com.inky.fitnesscalendar.db.entities.RichActivity
import com.inky.fitnesscalendar.db.entities.Track
import com.inky.fitnesscalendar.util.gpx.GpxReader
import java.io.FileInputStream
import java.io.InputStream
import javax.inject.Inject
import javax.inject.Singleton

@Immutable
@Singleton
class ImportRepository @Inject constructor(private val dbRepository: DatabaseRepository) {
    /**
     * Tries to import the track as activity and returns the activity id in case of success
     * or null if the track could not get imported
     */
    suspend fun importTrack(importTrack: ImportTrack, type: ActivityType): Int? {
        val activity = importTrack.toRichActivity(type) ?: return null
        val activityId = dbRepository.saveActivity(activity)
        val track = Track(activityId = activityId, points = importTrack.track.points)
        dbRepository.saveTrack(track)
        return activityId
    }

    /**
     * Recalculates statistics for all activities with a track, in case the calculation has changed
     * in an update, for example
     */
    suspend fun updateTrackActivities(): Int {
        var numChangedActivities = 0

        val tracks = dbRepository.loadTracks()
        for (track in tracks) {
            val richActivity = dbRepository.loadActivity(track.activityId)

            val updatedActivity = track.addStatsToActivity(richActivity.activity) ?: continue
            if (updatedActivity == richActivity.activity) {
                continue
            }
            numChangedActivities += 1
            dbRepository.saveActivity(richActivity.copy(activity = updatedActivity))
        }

        return numChangedActivities
    }

    fun loadFiles(files: List<ParcelFileDescriptor>): List<ImportTrack> {
        return files.flatMap {
            it.use { descriptor ->
                val actualDescriptor = descriptor.fileDescriptor
                val stream = FileInputStream(actualDescriptor)
                loadFile(stream)
            }
        }
    }


    private fun loadFile(stream: InputStream): List<ImportTrack> {
        val gpx = GpxReader.read(stream)
        return gpx?.tracks?.map { ImportTrack(track = it) } ?: emptyList()
    }

    data class ImportTrack(
        val track: GpxTrack,
        val dbTrack: Track = Track(activityId = -1, points = track.points),
        val trackSvg: TrackSvg? = track.toTrackSvg()
    ) {
        fun toRichActivity(type: ActivityType): RichActivity? {
            val initialActivity = Activity(
                typeId = type.uid ?: return null,
                description = track.name,
                startTime = track.startTime ?: return null,
                endTime = track.endTime ?: return null,
                trackPreview = trackSvg?.serialize()
            )

            val activity = dbTrack.addStatsToActivity(initialActivity) ?: return null

            return RichActivity(
                activity = activity,
                type = type,
                place = null,
            )
        }
    }
}