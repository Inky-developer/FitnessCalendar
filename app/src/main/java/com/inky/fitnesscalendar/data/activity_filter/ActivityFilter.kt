package com.inky.fitnesscalendar.data.activity_filter

import android.os.Parcelable
import com.inky.fitnesscalendar.data.ActivityCategory
import com.inky.fitnesscalendar.data.Vehicle
import com.inky.fitnesscalendar.db.entities.ActivityType
import com.inky.fitnesscalendar.db.entities.Place
import com.inky.fitnesscalendar.util.added
import kotlinx.parcelize.Parcelize


@Parcelize
data class ActivityFilter(
    val types: List<ActivityType> = emptyList(),
    val categories: List<ActivityCategory> = emptyList(),
    val places: List<Place> = emptyList(),
    val vehicles: List<Vehicle> = emptyList(),
    val text: String? = null,
    val range: DateRangeOption? = null,
    val attributes: AttributeFilter = AttributeFilter(),
) : Parcelable {
    fun isEmpty() = this == ActivityFilter()

    fun withCategory(newCategory: ActivityCategory) =
        copy(categories = categories.added(newCategory))

    fun withType(newType: ActivityType) = copy(types = types.added(newType))

    fun withPlace(newPlace: Place) = copy(places = places.added(newPlace))

    fun withVehicle(newVehicle: Vehicle) = copy(vehicles = vehicles.added(newVehicle))

    fun items(): List<ActivityFilterChip> {
        val items = mutableListOf<ActivityFilterChip>()

        if (!text.isNullOrBlank()) {
            items.add(ActivityFilterChip.TextFilterChip(text))
        }

        if (range != null) {
            items.add(ActivityFilterChip.DateFilterChip(range))
        }

        for (category in categories) {
            items.add(ActivityFilterChip.CategoryFilterChip(category))
        }

        for (type in types) {
            items.add(ActivityFilterChip.TypeFilterChip(type))
        }

        for (place in places) {
            items.add(ActivityFilterChip.PlaceFilterChip(place))
        }

        for (vehicle in vehicles) {
            items.add(ActivityFilterChip.VehicleFilterChip(vehicle))
        }

        for ((attribute, state) in attributes.entries()
            .filter { it.second != AttributeFilter.TriState.Undefined }) {
            items.add(
                ActivityFilterChip.AttributeFilterChip(
                    attribute,
                    state.toBooleanOrNull()!!
                )
            )
        }

        return items
    }
}