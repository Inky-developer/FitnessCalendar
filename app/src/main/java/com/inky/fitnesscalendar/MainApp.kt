package com.inky.fitnesscalendar

import android.app.Application
import android.util.Log
import androidx.core.net.toUri
import com.inky.fitnesscalendar.di.ActivityTypeOrder
import com.inky.fitnesscalendar.di.DecisionTrees
import com.inky.fitnesscalendar.repository.AppRepository
import com.inky.fitnesscalendar.util.cleanImageStorage
import com.inky.fitnesscalendar.util.getOrCreateImagesDir
import com.inky.fitnesscalendar.util.getOrCreateSharedMediaCache
import dagger.hilt.android.HiltAndroidApp
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltAndroidApp
class MainApp : Application() {
    @Inject
    lateinit var appRepository: AppRepository

    override fun onCreate() {
        super.onCreate()

        MainScope().launch(Dispatchers.IO) {
            initializeData()
            cleanupStorage()
        }
    }

    private suspend fun cleanupStorage() {
        val imageDir = getOrCreateImagesDir().toPath()
        val usedImages = appRepository.getUsedImages().map { imageDir.resolve(it).toFile().toUri() }
        cleanImageStorage(usedImages.toSet())

        getOrCreateSharedMediaCache().listFiles()?.forEach {
            it.delete()
        }
    }

    private suspend fun initializeData() {
        Log.i("MainApp", "Initializing app data")

        val activities = appRepository.loadMostRecentActivities(200)

        ActivityTypeOrder.init(activities)
        DecisionTrees.init(activities)
        Log.i("MainApp", "App data successfully initialized")
    }
}