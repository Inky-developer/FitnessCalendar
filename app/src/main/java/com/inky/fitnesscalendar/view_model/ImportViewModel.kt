package com.inky.fitnesscalendar.view_model

import android.content.Context
import android.content.Intent
import android.os.ParcelFileDescriptor
import androidx.compose.runtime.State
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.inky.fitnesscalendar.MainActivity
import com.inky.fitnesscalendar.R
import com.inky.fitnesscalendar.db.entities.ActivityType
import com.inky.fitnesscalendar.repository.DatabaseRepository
import com.inky.fitnesscalendar.repository.ImportRepository
import com.inky.fitnesscalendar.util.EXTRA_TOAST
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class ImportViewModel @Inject constructor(
    @ApplicationContext val context: Context,
    val dbRepository: DatabaseRepository,
    private val importRepository: ImportRepository
) : ViewModel() {
    var closeActivity: () -> Unit = {}

    private val _tracks = MutableStateFlow(emptyList<ImportRepository.ImportTrack>())
    val tracks get() = _tracks.asStateFlow()

    private val _done = mutableStateOf(false)
    val done: State<Boolean> get() = _done

    private val _error = mutableStateOf(false)
    val error: State<Boolean> get() = _error

    fun import(importTracks: List<Pair<ImportRepository.ImportTrack, ActivityType>>) =
        viewModelScope.launch(Dispatchers.Default) {
            var numImportedActivities = 0

            for ((importTrack, activityType) in importTracks) {
                val result = importRepository.importTrack(importTrack, activityType)
                if (result != null) {
                    numImportedActivities += 1
                }
            }

            val intent = Intent(context, MainActivity::class.java).apply {
                flags = Intent.FLAG_ACTIVITY_TASK_ON_HOME or Intent.FLAG_ACTIVITY_NEW_TASK
                putExtra(
                    EXTRA_TOAST,
                    context.resources.getQuantityString(
                        R.plurals.imported_activities,
                        numImportedActivities,
                        numImportedActivities
                    )
                )
            }
            context.startActivity(intent)
            closeActivity()
        }

    fun updateTypeMapping(key: String, value: ActivityType) =
        viewModelScope.launch(Dispatchers.IO) {
            dbRepository.setActivityTypeName(key, value)
        }

    fun loadFiles(files: List<ParcelFileDescriptor>) = viewModelScope.launch(Dispatchers.IO) {
        _done.value = false

        val loadedTracks = importRepository.loadFiles(files)
        if (loadedTracks.isEmpty()) {
            _error.value = true
        }

        _tracks.emit(_tracks.value.toMutableList().apply { addAll(loadedTracks) })

        _done.value = true
    }
}