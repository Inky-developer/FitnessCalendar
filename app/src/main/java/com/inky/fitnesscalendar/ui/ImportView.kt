package com.inky.fitnesscalendar.ui

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Add
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.CenterAlignedTopAppBar
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ExtendedFloatingActionButton
import androidx.compose.material3.FabPosition
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.input.nestedscroll.nestedScroll
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.inky.fitnesscalendar.R
import com.inky.fitnesscalendar.data.gpx.GpxTrack
import com.inky.fitnesscalendar.db.entities.Activity
import com.inky.fitnesscalendar.db.entities.ActivityType
import com.inky.fitnesscalendar.db.entities.RichActivity
import com.inky.fitnesscalendar.localization.LocalizationRepository
import com.inky.fitnesscalendar.ui.components.ActivityTypeSelector
import com.inky.fitnesscalendar.ui.components.BaseEditDialog
import com.inky.fitnesscalendar.ui.components.CompactActivityCard
import com.inky.fitnesscalendar.ui.util.localDatabaseValues
import com.inky.fitnesscalendar.view_model.ImportViewModel

@Composable
fun ImportView(viewModel: ImportViewModel) {
    val done by viewModel.done
    val tracks by viewModel.tracks.collectAsState()
    val error by viewModel.error

    if (!done) {
        CircularProgressIndicator(modifier = Modifier.fillMaxSize())
    } else if (error) {
        Scaffold {
            Text(
                stringResource(R.string.could_not_read_input_file),
                modifier = Modifier.padding(it)
            )
        }
    } else {
        ImportView(
            tracks = tracks,
            localizationRepository = viewModel.repository.localizationRepository,
            onImport = viewModel::import,
            onTypeMapping = viewModel::updateTypeMapping
        )
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ImportView(
    tracks: List<GpxTrack>,
    localizationRepository: LocalizationRepository,
    onImport: (List<RichActivity>) -> Unit,
    onTypeMapping: (String, ActivityType) -> Unit,
) {
    val typeMapping = localDatabaseValues.current.activityTypeNames
    val saveButtonEnabled = remember(tracks, typeMapping) {
        tracks.all {
            val type = typeMapping[it.type]
            type != null && it.toRichActivity(type) != null
        }
    }

    val scrollBehavior = TopAppBarDefaults.pinnedScrollBehavior()
    val topAppBarColors = TopAppBarDefaults.topAppBarColors(
        scrolledContainerColor = MaterialTheme.colorScheme.primaryContainer,
        titleContentColor = MaterialTheme.colorScheme.onPrimaryContainer
    )
    Scaffold(
        topBar = {
            CenterAlignedTopAppBar(
                title = { Text(stringResource(R.string.import_activities)) },
                colors = topAppBarColors,
                scrollBehavior = scrollBehavior,
            )
        },
        floatingActionButton = {
            AnimatedVisibility(visible = saveButtonEnabled) {
                ExtendedFloatingActionButton(
                    onClick = {
                        val activities = tracks.mapNotNull { track ->
                            typeMapping[track.type]?.let { track.toRichActivity(it) }
                        }
                        onImport(activities)
                    },
                ) {
                    Row(verticalAlignment = Alignment.CenterVertically) {
                        Icon(Icons.Outlined.Add, stringResource(R.string.confirm_import))
                        Text(
                            stringResource(R.string.confirm_import),
                            style = MaterialTheme.typography.titleMedium,
                            modifier = Modifier.padding(start = 8.dp)
                        )
                    }
                }
            }
        },
        floatingActionButtonPosition = FabPosition.Center,
        modifier = Modifier.nestedScroll(scrollBehavior.nestedScrollConnection)
    ) { paddingValues ->
        LazyColumn(
            contentPadding = PaddingValues(bottom = 128.dp),
            modifier = Modifier
                .padding(paddingValues)
                .fillMaxSize()
        ) {
            items(tracks) { track ->
                val type = typeMapping[track.type]
                TrackView(
                    track = track,
                    selectedType = type,
                    localizationRepository = localizationRepository,
                    onChangeType = { key, value -> onTypeMapping(key, value) }
                )
            }
        }
    }
}

@Composable
fun TrackView(
    track: GpxTrack,
    selectedType: ActivityType?,
    localizationRepository: LocalizationRepository,
    onChangeType: (String, ActivityType) -> Unit,
) {
    var dialogOpen by rememberSaveable { mutableStateOf(false) }
    var selectedActivityType by rememberSaveable { mutableStateOf<ActivityType?>(null) }
    val richActivity =
        remember(track, selectedType) { selectedType?.let { track.toRichActivity(selectedType) } }

    Box(modifier = Modifier.clickable { dialogOpen = true }) {
        if (richActivity != null) {
            CompactActivityCard(
                richActivity = richActivity,
                localizationRepository = localizationRepository,
                expand = true
            )
        } else {
            ActivityCardWithoutType(track, localizationRepository)
        }
    }

    if (dialogOpen) {
        BaseEditDialog(
            title = stringResource(R.string.choose_activity_type),
            onNavigateBack = { dialogOpen = false },
            onSave = {
                selectedActivityType?.let { onChangeType(track.type, it) }
                dialogOpen = false
            },
            actions = {}
        ) {
            ActivityTypeSelector(
                isSelected = { it == selectedActivityType },
                onSelect = { selectedActivityType = it }
            )
        }
    }
}

@Composable
private fun ActivityCardWithoutType(
    track: GpxTrack,
    localizationRepository: LocalizationRepository
) {
    val time = remember(track) {
        track.startTime?.let { localizationRepository.formatRelativeDate(it) } ?: ""
    }

    Card(
        colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surfaceContainer),
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 8.dp, vertical = 4.dp)
    ) {
        Text(
            time,
            style = MaterialTheme.typography.titleSmall,
            color = MaterialTheme.colorScheme.secondary,
            modifier = Modifier
                .align(Alignment.End)
                .padding(horizontal = 4.dp)
        )
        Text(
            track.type,
            style = MaterialTheme.typography.displaySmall,
            color = MaterialTheme.colorScheme.secondary,
            modifier = Modifier.padding(horizontal = 8.dp)
        )
        Text(
            track.name,
            overflow = TextOverflow.Ellipsis,
            maxLines = 1,
            color = MaterialTheme.colorScheme.secondary,
            style = MaterialTheme.typography.bodyMedium,
            modifier = Modifier.padding(horizontal = 8.dp)
        )
        Text(
            stringResource(R.string.click_to_set_type),
            color = MaterialTheme.colorScheme.primary,
            modifier = Modifier.padding(all = 8.dp)
        )
    }
}

private fun GpxTrack.toRichActivity(type: ActivityType): RichActivity? {
    return RichActivity(
        activity = Activity(
            typeId = type.uid ?: return null,
            description = name,
            startTime = startTime ?: return null,
            endTime = endTime ?: return null,
            distance = computeLength(),
        ),
        type = type,
        place = null
    )
}