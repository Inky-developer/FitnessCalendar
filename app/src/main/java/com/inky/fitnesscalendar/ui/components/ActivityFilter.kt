package com.inky.fitnesscalendar.ui.components

import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyItemScope
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Clear
import androidx.compose.material.icons.outlined.DateRange
import androidx.compose.material.icons.outlined.Edit
import androidx.compose.material3.Icon
import androidx.compose.material3.InputChip
import androidx.compose.material3.InputChipDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.SuggestionChip
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import com.inky.fitnesscalendar.R
import com.inky.fitnesscalendar.data.activity_filter.ActivityFilter
import com.inky.fitnesscalendar.data.activity_filter.ActivityFilterChip
import com.inky.fitnesscalendar.ui.util.SharedContentKey
import com.inky.fitnesscalendar.ui.util.localDatabaseValues
import com.inky.fitnesscalendar.ui.util.sharedElement

@Composable
fun FilterInformation(
    filter: ActivityFilter,
    onChange: (ActivityFilter) -> Unit
) {
    val context = LocalContext.current
    val listState = rememberLazyListState()
    val filterItems = remember(filter) { filter.items() }
    val historyItems = localDatabaseValues.current.activityFilterChips
    val filteredHistoryItems = remember(filter, historyItems) {
        historyItems.filter {
            !filterItems.contains(
                it
            )
        }
    }

    LaunchedEffect(filterItems) {
        listState.animateScrollToItem(0)
    }

    LazyRow(
        state = listState,
        modifier = Modifier
            .fillMaxWidth()
            .sharedElement(SharedContentKey.FilterInformation)
    ) {
        items(filterItems, key = { it }) { chip ->
            FilterChip(
                onClick = { onChange(chip.removeFrom(filter)) },
                label = { Text(chip.displayText(context)) },
                leadingIcon = { FilterChipIcon(chip) }
            )
        }

        items(filteredHistoryItems, key = { it }) { chip ->
            SuggestionFilterChip(
                onClick = { onChange(chip.addTo(filter)) },
                label = { Text(chip.displayText(context)) },
            )
        }
    }
}

@Composable
private fun FilterChipIcon(chip: ActivityFilterChip) {
    when (chip) {
        is ActivityFilterChip.AttributeFilterChip -> Icon(
            painterResource(R.drawable.outline_label_24),
            stringResource(R.string.attribute)
        )

        is ActivityFilterChip.CategoryFilterChip -> Text(
            chip.category.emoji,
            style = MaterialTheme.typography.titleLarge
        )

        is ActivityFilterChip.DateFilterChip -> Icon(
            Icons.Outlined.DateRange,
            stringResource(R.string.date_range)
        )

        is ActivityFilterChip.TextFilterChip -> Icon(
            Icons.Outlined.Edit,
            stringResource(R.string.text)
        )

        is ActivityFilterChip.TypeFilterChip -> Text(
            chip.type.emoji,
            style = MaterialTheme.typography.titleLarge
        )

        is ActivityFilterChip.PlaceFilterChip -> PlaceIcon(chip.place)

        is ActivityFilterChip.VehicleFilterChip -> Text(
            chip.vehicle.emoji,
            style = MaterialTheme.typography.titleLarge
        )

        is ActivityFilterChip.FeelFilterChip -> Text(
            chip.feel.emoji,
            style = MaterialTheme.typography.titleLarge
        )

        is ActivityFilterChip.FavoriteFilterChip -> FavoriteIcon(chip.favorite)
    }
}

@Composable
private fun LazyItemScope.FilterChip(
    leadingIcon: @Composable () -> Unit,
    label: @Composable () -> Unit,
    onClick: () -> Unit
) {
    InputChip(
        selected = false,
        onClick = onClick,
        label = label,
        leadingIcon = leadingIcon,
        trailingIcon = { Icon(Icons.Outlined.Clear, stringResource(R.string.clear)) },
        colors = InputChipDefaults.inputChipColors(containerColor = MaterialTheme.colorScheme.tertiaryContainer),
        modifier = Modifier
            .padding(horizontal = 4.dp)
            .animateItem()
    )
}

@Composable
private fun LazyItemScope.SuggestionFilterChip(
    label: @Composable () -> Unit,
    onClick: () -> Unit
) {
    SuggestionChip(
        onClick = onClick,
        label = label,
        icon = {
            Icon(
                painterResource(
                    R.drawable.outline_history_24,
                ),
                stringResource(R.string.recent)
            )
        },
        modifier = Modifier
            .padding(horizontal = 4.dp)
            .animateItem()
    )
}