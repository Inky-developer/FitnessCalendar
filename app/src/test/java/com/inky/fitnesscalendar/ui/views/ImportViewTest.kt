package com.inky.fitnesscalendar.ui.views

import android.os.Build
import androidx.compose.ui.test.ExperimentalTestApi
import androidx.compose.ui.test.assertCountEquals
import androidx.compose.ui.test.assertTextContains
import androidx.compose.ui.test.hasTestTag
import androidx.compose.ui.test.junit4.createComposeRule
import androidx.compose.ui.test.onAllNodesWithTag
import androidx.compose.ui.test.onNodeWithText
import androidx.compose.ui.test.performClick
import androidx.test.core.app.ApplicationProvider
import com.inky.fitnesscalendar.MainApp
import com.inky.fitnesscalendar.data.activity_filter.ActivityFilter
import com.inky.fitnesscalendar.repository.ImportRepository
import com.inky.fitnesscalendar.testUtils.mockDatabaseRepository
import com.inky.fitnesscalendar.ui.ImportView
import com.inky.fitnesscalendar.ui.util.ProvideDatabaseValues
import com.inky.fitnesscalendar.view_model.ImportViewModel
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.runBlocking
import org.approvaltests.Approvals
import org.approvaltests.utils.WithTimeZone
import org.junit.Assert.assertEquals
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import org.robolectric.shadows.ShadowLog


@Config(sdk = [Build.VERSION_CODES.UPSIDE_DOWN_CAKE])
@RunWith(RobolectricTestRunner::class)
class ImportViewTest {
    @get:Rule
    val composeTestRule = createComposeRule()

    /**
     * Test the import workflow:
     * - Load some tracks
     * - Go through each track and select an activity type (here biking)
     * - Verify that import errors are displayed
     * - Verify that the tracks were imported
     */
    @OptIn(ExperimentalTestApi::class)
    @Test
    fun test_can_import_activities() {
        val context = ApplicationProvider.getApplicationContext<MainApp>()

        val databaseRepository = mockDatabaseRepository(context)
        val viewModel = ImportViewModel(
            context = context,
            dbRepository = databaseRepository,
            importRepository = ImportRepository(databaseRepository)
        )

        val testFileNames = listOf(
            TestFile("/gpx/valid_track.gpx", false),
            TestFile("/gpx/invalid_track_no_track_points.gpx", true)
        )
        val testFiles = testFileNames.map { testFile ->
            ImportViewTest::class.java.getResource(testFile.path)!!.openStream()
                .use { it.readBytes().toString(Charsets.UTF_8) }
        }

        runBlocking {
            viewModel.loadTestFiles(testFiles)
        }

        assertEquals("Both tracks should have been imported", 2, viewModel.tracks.value.size)

        composeTestRule.setContent {
            ProvideDatabaseValues(databaseRepository) {
                ImportView(viewModel)
            }
        }

        ShadowLog.stream = System.out

        composeTestRule.apply {
            waitUntilExactlyOneExists(hasTestTag("import_list"))

            val listItems = onAllNodesWithTag("import_list_item").assertCountEquals(2)
            for ((index, _) in listItems.fetchSemanticsNodes().withIndex()) {
                val node = listItems[index]

                // Select the activity type for biking for this activity
                node.performClick()
                onNodeWithText("ðŸš´").performClick()
                onNodeWithText("Save").performClick()
                waitForIdle()

                if (testFileNames[index].isError) {
                    node.assertTextContains("An error occurred.", substring = true)
                } else {
                    node.assertTextContains("ðŸš´", substring = true)
                }
            }

            runBlocking {
                assertEquals(
                    "Initially, there should be no activities",
                    0,
                    databaseRepository.getActivities(ActivityFilter()).first().size
                )
            }

            onNodeWithText("Import").performClick()

            waitUntil(timeoutMillis = 5000) {
                runBlocking {
                    databaseRepository.getActivities(ActivityFilter()).first().isNotEmpty()
                }
            }
        }
        runBlocking {
            WithTimeZone("UTC").use {
                Approvals.verifyAll(
                    databaseRepository.getActivities(ActivityFilter()).first().toTypedArray()
                ) { it.toString() }
            }
        }

    }

    data class TestFile(val path: String, val isError: Boolean)
}