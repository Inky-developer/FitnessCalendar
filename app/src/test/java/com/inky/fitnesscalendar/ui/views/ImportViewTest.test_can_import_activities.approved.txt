RichActivity(activity=Activity(uid=1, typeId=1, placeId=null, vehicle=null, description=2024-12-18T20:26+01, favorite=false, imageName=null, feel=Ok, intensity=null, startTime=Wed Dec 18 19:27:02 UTC 2024, endTime=Wed Dec 18 19:27:02 UTC 2024, distance=Distance(meters=0.0), movingDuration=Duration(elapsedMs=0), temperature=null, averageHeartRate=null, maximalHeartRate=null, totalAscent=Distance(meters=0.0), totalDescent=Distance(meters=0.0), trackPreview=SerializedTrackPreview(inner={"points":[{"latitude":37.421998,"longitude":-122.084}],"bounds":{"latitudeMin":37.421998,"latitudeMax":37.421998,"longitudeMin":-122.084,"longitudeMax":-122.084}})), type=Cycling, place=null)
